services/supervisor/src/lib.rs

```diff
@@ -34,19 +29,72 @@ use exonum::{
     helpers::{byzantine_quorum, validator::validator_id},
     runtime::{
         api::ServiceApiBuilder,
-        rust::{AfterCommitContext, BeforeCommitContext, Service, Transaction},
+        rust::{AfterCommitContext, CallContext, Service, Transaction},
         InstanceDescriptor, SUPERVISOR_INSTANCE_ID, SUPERVISOR_INSTANCE_NAME,
     },
 };
+use exonum_derive::*;
 use exonum_merkledb::Snapshot;
 
 mod api;
+mod configure;
 mod errors;
 mod proto;
 mod proto_structures;
 mod schema;
+pub mod simple;
 mod transactions;
 
+/// Applies configuration changes, isolating each of them with by using `Fork` checkpoints.
+///
+/// # Safety
+///
+/// This function should be used with extreme care. It makes the following assumptions:
+///
+/// - The function must be called at the end of the transaction execution. If the transaction
+///   errors / panics afterwards, the changes to the configs will not be rolled back.
+/// - No changes to the blockchain state should be introduced before the call to this function.
+///   Any changes that are introduced will be committed regardless of the execution status,
+///   or the status of application of any config change. This is if the execution wasn't interrupted
+///   by a panic / error *before* hitting the call; if this happens, the usual rules apply.
+///
+/// These restrictions are the result of `Fork` not having multi-layered checkpoints.
+fn update_configs(context: &mut CallContext, changes: Vec<ConfigChange>) {
+    // An error while configuring one of the service instances should not affect others.
+    changes.into_iter().for_each(|change| match change {
+        ConfigChange::Consensus(config) => {
+            log::trace!("Updating consensus configuration {:?}", config);
+
+            let result = context.isolate(|context| {
+                blockchain::Schema::new(context.fork())
+                    .consensus_config_entry()
+                    .set(config);
+                Ok(())
+            });
+            assert!(result.is_ok());
+        }
+
+        ConfigChange::Service(config) => {
+            log::trace!(
+                "Updating service instance configuration, instance ID is {}",
+                config.instance_id
+            );
+
+            let configure_result = context.isolate(|mut context| {
+                context
+                    .interface::<ConfigureCall>(config.instance_id)?
```

Comment:
> every service

Every service *that wants to be configured via supervisor*, yes. I don't see why this is a problem, but it would be easy to move `Configure`-related stuff to a separate crate.

Reply:
>I don't see why this is a problem
I think it can become a problem on a cross-runtime border: e.g. configurable Java service with a Rust supervisor.

So yeah, IMHO it's better to keep the interfaces separate from the implementation.

Reply:
Can we defer this question for later?

Reply:
Sure.